ibtk/include/ibtk/SnapshotCache.h

```diff
@@ -0,0 +1,285 @@
+    }
+
+    /*!
+     * Get a const iterator to element following the last element of the snapshots.
+     *
+     * Dereferencing this element results in undefined behavior.
+     */
+    inline const_iterator end() const
+    {
+        return d_snapshots.end();
+    }
+
+    /*!
+     * Get a const iterator to the first element of the snapshots.
+     */
+    inline iterator begin()
+    {
+        return d_snapshots.begin();
+    }
+
+    /*!
+     * Get an iterator to element following the last element of the snapshots.
+     *
+     * Dereferencing this element results in undefined behavior.
+     */
+    inline iterator end()
+    {
+        return d_snapshots.end();
+    }
+
+    /*!
+     * Get the number of stored snapshots.
+     */
+    inline size_t getNumSnapshots() const
+    {
+        return d_snapshots.size();
+    }
+
+    /*!
+     * Write the snapshot cache to a database. This writes both the time points and the snapshotted patch hierarchies.
+     */
+    void putToDatabase(SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> db) override;
+
+    /*!
+     * Get the patch index owned by this class.
+     */
+    inline int getPatchIndex() const
+    {
+        return d_snapshot_idx;
+    }
+
+    /*!
+     * Get the variable owned by this class.
+     */
+    inline SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > getVariable() const
+    {
+        return d_snapshot_var;
+    }
+
+private:
+    /*!
+     * Read stored information from the restart files. Note the GridGeometry object is required in order to construct
+     * the snapshotted patch hierarchies.
+     */
+    void getFromRestart(SAMRAI::tbox::Pointer<SAMRAI::hier::GridGeometry<NDIM> > grid_geom);
+
+    /*!
+     * \brief Default constructor.
+     *
+     * \note This constructor is not implemented and should not be used.
+     */
+    SnapshotCache() = delete;
+
+    /*!
+     * \brief Copy constructor.
+     *
+     * \note This constructor is not implemented and should not be used.
+     *
+     * \param from The value to copy to this object.
+     */
+    SnapshotCache(const SnapshotCache& from) = delete;
+
+    /*!
+     * \brief Assignment operator.
+     *
+     * \note This operator is not implemented and should not be used.
+     *
+     * \param that The value to assign to this object.
+     *
+     * \return A reference to this object.
+     */
+    SnapshotCache& operator=(const SnapshotCache& that) = delete;
+
+    std::string d_object_name;
+
+    /*
+     * Data for tracking snapshots.
+     */
+    SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > d_snapshot_var;
+    SAMRAI::tbox::Pointer<SAMRAI::hier::VariableContext> d_ctx;
+    int d_snapshot_idx = IBTK::invalid_index;
+    SAMRAI::hier::IntVector<NDIM> d_gcw = 1;
+
+    /*!
+     * Vector of snapshots consisting of time points and hierarchies. Each hierarchy allocates data only with the patch
+     * data index stored by this class.
+     */
+    std::vector<SnapshotType> d_snapshots;
```

Comment:
If we are returning iterators anywhere then this is not going to work since when `std::vector` grows it must allocate a completely new array. If we can guarantee that we never delete snapshots in the middle of the range then either `std::deque<T>` or `std::vector<std::unique_ptr<T>>` is the right choice for avoiding this problem.

Reply:
The storage type is a double and a pointer. Resizing a vector will be very cheap already. If you're concerned about iterator invalidation, `std::deque` will have also have iterators invalidated when an element is inserted. I'm not sure I understand the benefit of changing to `deque`.

Reply:
Resizing on its own is definitely not a problem, just the resulting iterator invalidation.

The point of changing to `std::deque` is that if we never need to insert a snapshot into the middle (i.e., we only a ever append or prepend) then these iterators are always valid. I imagine that, in practice, this is always what we will do.

If we want to support that case then we really do need to not return iterators and use `std::vector<std::unique_ptr<T>>` since the iterators will be invalidated every time we add enough snapshots. That combination (no iterators and unique pointers) will always work.

Reply:
that's a good idea - lets put that in `ibtk_utilities.h`.

We could also require C++14, but sometimes people still want to build IBAMR with GCC 4.x, so lets do C++11 for at least one more release.